### ISSUE TITLE
Unable to apply filters for Redis connection

### ISSUE THREAD

**Issue by bocharsky-bw** [TablePlus/TablePlus#1094]
1. Which driver are you using and version of it: Redis 4.0.10

2. Which TablePlus build number are you using: the latest for Mac currently - 212

3. The steps to reproduce this issue:

I use "key Contains" filter but the output is always weird no matter what to search - it just show "value" column and a single number in it
<img width="1143" alt="Screen Shot 2019-04-09 at 10 02 53" src="https://user-images.githubusercontent.com/3317635/55779648-aaed5280-5aae-11e9-867d-eb736da44c40.png">

**Comment by huyphams** [comment]
Thank @bocharsky-bw , have you tried to remove the double quote?

**Comment by huyphams** [comment]
<img width="1194" alt="Screen Shot 2019-04-09 at 2 05 48 PM" src="https://user-images.githubusercontent.com/3654622/55779783-9621b680-5ad0-11e9-8fc0-e6af3b2ad5bc.png">

**Comment by bocharsky-bw** [comment]
Yes, I did! Nothing happen, the same output

**Comment by bocharsky-bw** [comment]
Hm, probably the issue in Redis version? I see you use 5.0.0 :/

**Comment by huyphams** [comment]
No, it should work for all version. Are you sure that you have a key which contains test?

**Comment by huyphams** [comment]
This is Redis 3
<img width="1194" alt="Screen Shot 2019-04-09 at 2 12 11 PM" src="https://user-images.githubusercontent.com/3654622/55780147-7b037680-5ad1-11e9-92d2-179e26b9fd6f.png">

**Comment by bocharsky-bw** [comment]
Ah, I think I know the problem, the filter applied only on the current (loaded) page. For example, if I go the 2nd page below with pagination, copy a key from there, paste it into Pattern field and try to search by it - it won't be found because filtering was applied on the 1st page. Then, if I moved to the 2nd page with paginator below - it will find my key!

But is it desired behavior? I thought filtering should work for the whole data, not only loaded data for current page, as it works for MySQL connection

**Comment by huyphams** [comment]
I don't think it works that way, but you said that so I'm confused now.
TablePlus generate the command `SCAN 0 MATCH *a* COUNT 35` so if you go to page 2, it will be `SCAN 300 MATCH *a* COUNT 35`, so basically, it will apply the filter for the whole database, just like MySQL.

**Comment by huyphams** [comment]
Could you please click on the SQL button below the filter bar, copy the generated command and run it directly in Query Editor? I would like to see the result, maybe the bug is there thank you @bocharsky-bw

**Comment by bocharsky-bw** [comment]
Hm, first of all I have a few indexes (Items in TablePlus): `db0` and `db1`. I want to search in `db1` item, but SQL button shows me:
```
-- Current filter --
SCAN 0 MATCH *test* COUNT 10
```

Shouldn't it show me `SCAN 0 MATCH db1:*test* COUNT 10` or something like this?

And to reproduce my problem, you just need to have a few pages. You can go to the 2nd page and copy any key from there, then put it into Pattern and press Enter. It won't find the key... but if you then just click on the next page button - it finds the key. Though, SQL button still shows me:
```
-- Current filter --
SCAN 0 MATCH *test* COUNT 10
```

What is "Query Editor"? Is it "Console" tab when I press on `SQL` icon left sidebar? Anyway, I think you should be able to reproduce the problem, just set limit to 10 or 5 or less so you have a few pages and try to find a key that is on the next page, on the 1st one.

**Comment by huyphams** [comment]
`Hm, first of all I have a few indexes (Items in TablePlus): db0 and db1. I want to search in db1 item, but SQL button shows me:` => TablePlus always run SELECT db_index in order to run the SCAN statement so it will be no problem.

`just set limit to 10 or 5 or less so you have a few pages and try to find a key that is on the next page, on the 1st one.` I think the issue when the range is out, Redis will return the count value instead of the key, I got value 3 when I ran `SCAN 20 MATCH *b* COUNT 5` I will fix it, it should be empty.

**Comment by bocharsky-bw** [comment]
Hey @huyphams !

So, I think I finally got how it works, at least how it works in TablePlus. Unfortunately, I am far from an expert in Redis, but I noticed that because of limit (that is set to 35 by default) Filter applied to the first  35 rows, and this is confirmed by SQL button - when you press SQL you see that behind the scene it executes the next query:
```
-- Current filter --
SCAN 0 MATCH *test* COUNT 35
```

So, if you have a lot of entries (as I do) like about 50k rows - TablePlus will search test key in first 35 rows only. I also found a workaround: after you applied a filter, you can go to the ⚙️ icon and set the limit manually to 50000 and apply the filter one more time. And this time it will search `*test*` in the whole table, i.e. not only in first 35 rows.

And this workaround works fine except one problem: when you remove Filter and forget to revert limit back to 35 - it will try to load all those 50k rows on one page... and that's a problem of course because it just hangs.

As a possible solution, I think you should use the total number of rows instead of limit in the query behind the scene, i.e. as soon as I apply a Filter, execute:
```
-- Current filter --
SCAN 0 MATCH *test* COUNT %total_rows%
```

Where `%total_rows%` is 50000 in my case, but sure it should be dynamic.

Maybe there's another possible solution because as I said I'm far from Redis expert yet :)

Does it sound clear for you?
P.S. thanks on your quick help on it btw!

**Comment by huyphams** [comment]
Well, I didn't know that thank you for pointing it out, I had no idea why I missed it when I read the document...

**Comment by bocharsky-bw** [comment]
No problem! I even don't know if it's described well in docs... I just figure it out at random today. Anyway, thank you for supporting Redis in TablePlus, that's awesome!

**Comment by huyphams** [comment]
I did a quick fix: https://www.dropbox.com/s/0i0ofnhi88vlxz8/TablePlus.dmg?dl=0
TablePlus now will scan all keys, but I'm not sure if it costs performance. Let me know if you have any issue.

**Comment by bocharsky-bw** [comment]
Yeah, probably costs, but an ability to search in the whole DB is a good DX I think because it's the same behavior as with MySQL DB, etc. FYI, search by ~50k took me about 1-1.5 seconds only, so it's almost unnoticed to search in 1000 rows or 50k rows.

Just checked, works great except one minor detail: when I remove text in the Pattern field and press enter - it executes this query:
```
-- Current filter --
SCAN 0 MATCH ** COUNT %total_rows%
```

and hangs... so it would be good to ignore empty string, i.e. if an empty string in the Pattern - use limit - the same behavior as if you just disable a filter:
```
-- Current filter --
SCAN 0 COUNT %limit%
```

**Comment by bocharsky-bw** [comment]
Oh, and one more problem noticed! When results are many, it ignores pagination :/ I mean, the limit is set to 35 rows, below I see `1-35 of 58,098 rows`, but on the page, I see all the results - all 48 rows that were found. Though, not sure how to add pagination to `SCAN 0 MATCH ** COUNT %total_rows%` query, i.e. search by all entries but respect `limit` and display only some of them

**Comment by huyphams** [comment]
Thank for the feedback, I ignored the space: https://www.dropbox.com/s/0i0ofnhi88vlxz8/TablePlus.dmg?dl=0
I still can't figure out how to limit the SCAN MATCH pattern, the COUNT is using for the "amount of work" not for the limitation in this case.

**Comment by bocharsky-bw** [comment]
Thanks! Yeah, don't know how to work around it :/
